backend/app/categories/controller.py

```diff
@@ -80,5 +83,52 @@ async def rate_item_route(category_id: CategoryEnum, item_id: ObjectID, rating:
                           token: dict = Depends(verify_token),
                           db: AsyncIOMotorClient = Depends(get_database)) -> ResponseBase:
     """"Rate a given item"""
-    # TODO add rating functionality here
+
+    # return success false if item is invalid
+    item_test = await db[category_id]["data"].find_one({'_id': item_id})
+    if item_test == None:
+        return {"success": False}
+
+    # get user
+    user_name = token["username"]
+    user = None
+    async for item in db.core.users.find():
+        if item["username"] == user_name:
+            user = item
+            break
+
+    # return success false if user is invalid
+    if user == None:
+        return {"success": False}
+
+    user_id = user["_id"]
+    rating_list = user["ratings"]
+
+    # create Rating object
+    rating_here = Rating(user_id=user_id, rating=rating, item_id=item_id)
+
+    # ensure one rating per person per item
+    existing = await db[category_id]["ratings"].find_one({'item_id': item_id, 'user_id': user_id})
```

Comment:
Instead of checking if it exists separately, you can just use the `upsert: true` [flag](https://docs.mongodb.com/manual/reference/method/db.collection.update/#update-upsert) on the `update_one` function.

```python
_res = await db[category_id]["ratings"].update_one({ 'item_id': item_id, 'user_id': user_id}, {
    "$set": {"rating": rating},
}, {"upsert": true})
rating_id = None
if _res.upserted:
    rating_id = _res.upserted
else:
    rating_id = _res._id